cli/src/main/java/org/eclipse/hono/cli/AbstractCliClient.java

```diff
@@ -44,32 +59,87 @@
+    /**
+     * Spring Shell inputReader.
+     */
+    @Autowired
+    protected @Lazy InputReader inputReader;
+    /**
+     * Spring Shell output helper.
+     * <p>
+     * it can be passed to the method class instead to use the log (it's more stylish).
+     */
+    @Autowired
+    protected ShellHelper shellHelper;
+    /**
+     * The client factory methods.
+     */
+    private ConnectionLifecycle<HonoConnection> clientFactory;
+    /**
+     * The active profile.
+     */
+    private Profile profile;
     /**
-     * Sets the Spring environment.
-     * 
-     * @param env The environment.
-     * @throws NullPointerException if environment is {@code null}.
+     * Sets the default client configuration.
+     *
+     * @param honoClientConfig The Client Configuration class instance.
      */
     @Autowired
-    public final void setActiveProfiles(final Environment env) {
-        Objects.requireNonNull(env);
-        activeProfiles = Arrays.asList(env.getActiveProfiles());
+    private void setClientConfigProperties(final ClientConfigProperties honoClientConfig) {
+        this.honoClientConfig = Objects.requireNonNull(honoClientConfig);
     }
-
     /**
      * Sets the vert.x instance.
-     * 
+     *
      * @param vertx The vert.x instance.
      * @throws NullPointerException if vert.x is {@code null}.
      */
     @Autowired
-    public final void setVertx(final Vertx vertx) {
+    private void setVertx(final Vertx vertx) {
         this.vertx = Objects.requireNonNull(vertx);
         this.ctx = vertx.getOrCreateContext();
     }
+    /**
+     * Method to set the profile and create the factory for the application profile.
+     *
+     * @param honoClientConfig the client configuration parameters.
+     * @param profile the profile as the connection will be opened.
+     */
+    protected void createFactory(final ClientConfigProperties honoClientConfig, final Profile profile) {
+        this.profile = profile;
+        if (profile == Profile.APPLICATION) {
+            clientFactory = ApplicationClientFactory.create(HonoConnection.newConnection(vertx, honoClientConfig));
+        } else if (profile == Profile.DEVICE) {
+            clientFactory = AmqpAdapterClientFactory.create(HonoConnection.newConnection(vertx, honoClientConfig), honoClientConfig.getUsername().split("@")[1]);
+        }
+    }
+    /**
+     * Accessor to destroy the factory and reset the profile.
+     *
+     */
+    protected void destroyFactory() {
+        profile = null;
+        clientFactory = null;
+    }
+    /**
+     * Accessor to the application's methods factory.
+     *
+     * @return the factory
+     */
+    protected ConnectionLifecycle<HonoConnection> getClientFactory() {
+        return StringUtils.isEmpty(profile) ? null : clientFactory;
+    }
+    /**
+     * Accessor to the active profile.
+     *
+     * @return The active profile.
+     */
+    protected Profile getProfile() {
+        return profile;
+    }
```

Comment:
none of these methods seem to be used in `AbstractCliClient` but their implementation seems to be specific to the concrete subclasses. So FMPOV these methods should be removed here and should be implemented in the concrete subclasses instead. There is no need to duplicate the fact that we have multiple subclasses by introducing the `Profile` enum whose sole purpose seems to be to indicate, in which subclass we are currently running code ...

Reply:
Can i delete the disconnect command?(And the class Client with it)

Removing the profile flag and moving down the factory, that method needs something like a boolean flag in the `AbstractCliClient` class so i can interrupt the availability of the single profile when the disconnect is called, but i don't like the flag thing, and also from the User POV calling disconnect or calling another `connect-something` doesn't change anything.

However (just to underline it) moving the factory down of a level and removing the disconnect, if he user call a

> connect-application

and then a 

> connect-device

the memory remains allocated for both factories. Not a big deal, but the only solution to this is to make two disconnect methods that FMPOV are evil for a user.

Reply:
Keeping the factory instances is no problem at all. I don't quite get why having two disconnect methods would be a bad thing. What other way would there be for a user to indicate whether he wants to close the connection to the north bound API or the protocol adapter? There are also two different methods for establishing the connection aren't there?

Reply:
Actually the `disconnect` and `connect` not disconnect or connect just create the assumptions to execute the commands relative to the profile choose. Only at the execution of the command the connection is instantiated.
So why bother the user to choose what he want to disconnect? and furthermore what's the difference for him if he disconnect or not? (I reasoned in the same way for the connect but you said that for future implementation like MQTT adapter was better to make it specific). 
(I said 'evil' for the user not 'bad' 😊)
However if you say two disconnect, i'll do so.


(update)
I can't disconnect (i.e. `clientFactory = null`) from another class, the variable it's not in common between the class Client and the others 2 of the profiles, and that's why i used previously the bean class ClientConfig. So i guess i'll go for 2 disconnect beside what i have said.